src/librbd/cache/ParentCacheObjectDispatch.cc

```diff
@@ -0,0 +1,234 @@
+    }
+
+    ldout(cct, 5) << "session don't work, dispatch current request to lower object layer " << dendl;
+    return false;
+  }
+  ceph_assert(m_cache_client->is_session_work());
+
+  CacheGenContextURef ctx = make_gen_lambda_context<ObjectCacheRequest*,
+                                     std::function<void(ObjectCacheRequest*)>>
+   ([this, snap_id, read_data, dispatch_result, on_dispatched,
+      oid, object_off, object_len](ObjectCacheRequest* ack) {
+     handle_read_cache(ack, object_off, object_len, read_data,
+                       dispatch_result, on_dispatched);
+  });
+
+  m_cache_client->lookup_object(m_image_ctx->data_ctx.get_namespace(),
+                                m_image_ctx->data_ctx.get_id(),
+                                (uint64_t)snap_id, oid, std::move(ctx));
+  return true;
+}
+
+template <typename I>
+void ParentCacheObjectDispatch<I>::handle_read_cache(
+     ObjectCacheRequest* ack, uint64_t read_off, uint64_t read_len,
+     ceph::bufferlist* read_data, io::DispatchResult* dispatch_result,
+     Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  if(ack->type != RBDSC_READ_REPLY) {
+    // go back to read rados
+    *dispatch_result = io::DISPATCH_RESULT_CONTINUE;
+    on_dispatched->complete(0);
+    return;
+  }
+
+  ceph_assert(ack->type == RBDSC_READ_REPLY);
+  std::string file_path = ((ObjectCacheReadReplyData*)ack)->cache_path;
+  ceph_assert(file_path != "");
+
+  // try to read from parent image cache
+  int r = m_object_store->read_object(file_path, read_data, read_off, read_len, on_dispatched);
+  if(r < 0) {
+    // cache read error, fall back to read rados
+    *dispatch_result = io::DISPATCH_RESULT_CONTINUE;
+    on_dispatched->complete(0);
+  }
+
+  *dispatch_result = io::DISPATCH_RESULT_COMPLETE;
+  on_dispatched->complete(r);
+}
+
+template <typename I>
+int ParentCacheObjectDispatch<I>::handle_register_client(bool reg) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  if (reg) {
+    ldout(cct, 20) << "Parent cache open cache handler" << dendl;
+    m_object_store = new SharedPersistentObjectCacher(m_image_ctx);
+  }
+  return 0;
+}
+
+template <typename I>
+int ParentCacheObjectDispatch<I>::create_cache_session(Context* on_finish, bool is_reconnect) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  Context* register_ctx = new FunctionContext([this, cct, on_finish](int ret) {
+    if (ret < 0) {
+      lderr(cct) << "Parent cache fail to register client." << dendl;
+      handle_register_client(false);
+      on_finish->complete(-1);
+      return;
+    }
+    ceph_assert(m_cache_client->is_session_work());
+
+    handle_register_client(true);
+    on_finish->complete(0);
+  });
+
+  Context* connect_ctx = new FunctionContext(
+    [this, cct, register_ctx](int ret) {
+    if (ret < 0) {
+      lderr(cct) << "Parent cache fail to connect RO daeomn." << dendl;
+      register_ctx->complete(-1);
```

Comment:
Nit: just pass the existing ```ret``` instead of hard-coding ```-1```

Reply:
done. i have refined this context implements. thanks